src/client/datascience/jupyter/kernels/kernelExecution.ts

```diff
@@ -57,36 +55,35 @@ export class KernelExecution implements IDisposable {
     }
 
     @captureTelemetry(Telemetry.ExecuteNativeCell, undefined, true)
-    public async executeCell(cell: NotebookCell): Promise<void> {
-        if (!this.notebook) {
-            throw new Error('executeObservable cannot be called if kernel has not been started!');
-        }
+    public async executeCell(notebookPromise: Promise<INotebook>, cell: NotebookCell): Promise<void> {
         // Cannot execute empty cells.
         if (this.cellExecutions.has(cell) || cell.document.getText().trim().length === 0) {
             // clear cell output
             cell.outputs = [];
             return;
         }
+        const editor = this.vscNotebook.notebookEditors.find((item) => item.document === cell.notebook);
+        if (!editor) {
+            return;
+        }
         const cellExecution = this.executionFactory.create(cell, isPythonKernelConnection(this.metadata));
         this.cellExecutions.set(cell, cellExecution);
 
-        const kernel = this.getKernel(cell.notebook);
-
         try {
+            const notebook = await notebookPromise;
```

Comment:
At this point, the status of the cell has been updated & we wait for the notebook.
In the `finally` we call `cellExecution.cancel` which reverts the cell state.

Reply:
`cellExecution` sets the state when its created & udpates again when it is disposed/cancelled/completed.